spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/BaseCopyTableSparkAction.java

```diff
@@ -0,0 +1,871 @@
+      }
+    }
+  }
+
+  private boolean versionInFilePath(String path, String version) {
+    return fileName(path).equals(version);
+  }
+
+  private String jobDesc() {
+    if (startVersion.isEmpty()) {
+      return String.format(
+          "Replacing path prefixes '%s' with '%s' in the metadata files of table %s,"
+              + "up to version '%s'.",
+          sourcePrefix, targetPrefix, table.name(), endVersion);
+    } else {
+      return String.format(
+          "Replacing path prefixes '%s' with '%s' in the metadata files of table %s,"
+              + "from version '%s' to '%s'.",
+          sourcePrefix, targetPrefix, table.name(), startVersion, endVersion);
+    }
+  }
+
+  /**
+   * Here are steps: 1. rebuild version files 2. rebuild manifest list files 3. rebuild manifest
+   * files 4. get all data files need to move
+   */
+  private void rebuildMetaData() {
+    TableMetadata tableMetadata = ((HasTableOperations) endStaticTable).operations().current();
+
+    // rebuild version files
+    Set<Long> allSnapshotIds = rewriteVersionFiles(tableMetadata);
+
+    Set<Long> diffSnapshotIds = getDiffSnapshotIds(allSnapshotIds);
+
+    // get all manifest file paths need to rewrite
+    List<String> manifestFilePathToMove = manifestFilesToMove(diffSnapshotIds);
+    manifestFilePaths.addAll(manifestFilePathToMove);
+
+    // rebuild manifest-list files
+    Set<Snapshot> validSnapshots =
+        Sets.difference(snapshotSet(endVersion), snapshotSet(startVersion));
+    validSnapshots.forEach(snapshot -> rewriteManifestList(snapshot, tableMetadata));
+
+    // rebuild manifest files
+    List<ManifestFile> newManifests = rewriteManifests(tableMetadata);
+    newManifests.forEach(manifestFile -> addToRebuiltFiles(manifestFile.path()));
+    saveMetadataFileList();
+
+    Dataset<Row> dataFiles = getDiffDataFiles(diffSnapshotIds);
+    saveDataFileList(dataFiles);
+  }
+
+  private void saveMetadataFileList() {
+    List<String> fileList = Lists.newArrayList();
+    fileList.addAll(metadataFilesToMove);
+    Dataset<String> metadataFileList = spark().createDataset(fileList, Encoders.STRING());
+    metadataFileListPath = stagingDir + METADATA_FILE_LIST_DIR;
+    metadataFileList
+        .repartition(1)
+        .write()
+        .mode(SaveMode.Overwrite)
+        .format("text")
+        .save(metadataFileListPath);
+  }
+
+  private void saveDataFileList(Dataset<Row> dataFiles) {
+    dataFileListPath = stagingDir + DATA_FILE_LIST_DIR;
+
+    try {
+      dataFiles
+          .repartition(1)
+          .write()
+          .mode(SaveMode.Overwrite)
+          .format("text")
+          .save(dataFileListPath);
+    } catch (Exception e) {
+      throw new UnsupportedOperationException(
+          "Failed to build the data files dataframe, the end version you are "
+              + "trying to copy may contain invalid snapshots, please use the younger version which doesn't have invalid "
+              + "snapshots",
+          e);
```

Comment:
If we want to support actually moving the files - then we will need to support different cloud providers (GCP, AWS, Azure) as well as on-prem setups (i.e copying in Linux). It might not be ideal for some use-cases as well - for example in our case where we are actually using this in a production setting, copying using GCS client libraries was extremely slow and inefficient for huge tables (>5TB) and instead we used a [managed service from GCP](https://cloud.google.com/storage-transfer/docs/cloud-storage-to-cloud-storage ) to handle the move efficiently. So this can actually be very specific to where the files are stored and where you want to move them, you can consider a use-case of someone migrating between 2 different cloud providers or moving data from on-premise to the cloud for example and so on. So from my perspective for it's better to just rewrite the paths so the table is usable in a new location and leave the actual copying to the users, maybe in a future iteration some basic interface to move files for common use-cases can be implemented